python-sgd-regression/sgd_regression.py

```diff
@@ -0,0 +1,149 @@
+    for k, table in enumerate(tables):
+        # Load partial model
+        if partial_job_id:
+            estimator = _load_partial_estimator(partial_job_id)
+        else:
+            # TODO: add optional parameters to `SGDRegressor`
+            estimator = SGDRegressor()
+
+        # TODO: normalize estimator
+
+        # TODO: PARAM_query can only handle single table -> we need to generalize this somehow
+        query = io_helper._get_query()
+        query_table = query.replace('FROM SAMPLE_DATA', 'FROM {}'.format(table))
+
+        # TODO: `fetch_data` needs `sql` as optional parameter instead of hardcoded `_get_query`
+        inputs = io_helper.fetch_data(sql=query_table)
+
+        # TODO: following is copied from python-linear-regression -> DRY it and put into common library (either
+        # io_helpers or a new util library in `algorithm-repository`)
+        dep_var = inputs["data"]["dependent"][0]
+        inped_vars = inputs["data"]["independent"]
+
+        # Check dependent variable type (should be continuous)
+        if dep_var["type"]["name"] not in ["integer", "real"]:
+            logging.warning("Dependent variable should be continuous !")
+            return None
+
+        # Extract data and parameters from inputs
+        data = format_data(inputs["data"])
+
+        # Select dependent and independent variables with patsy
+        X, y = get_Xy(dep_var, inped_vars, data)
+
+        # Train single step
+        estimator.partial_fit(X, y)
+
+        # Save intermediate serialized model into job_result table
+        # TODO: skip this step if this is the last iteration
+        estimator_json = serialize_sklearn_estimator(estimator)
+
+        # Create new job with job_id in format `[job_id]-[iteration]` for storing partial results
+        partial_job_id = '{}-{}'.format(job_id, k)
+
+        # TODO: `save_results` need `job_id` as optional argument
+        io_helper.save_results(estimator_json, '', 'json', job_id=partial_job_id)
+
+    # Generate PFA output from the final model and save it
+    pfa = sklearn_to_pfa(estimator)
+    io_helper.save_results(pfa, '', 'pfa_json')
+
+
+def _load_partial_estimator(job_id):
+    engine = create_engine(io_helper._get_output_jdbc_url())
+    session = sessionmaker(bind=engine)()
+
+    job_result = session.query(JobResult).filter_by(job_id=job_id).first()
+    deserialize_sklearn_estimator(job_result.data)
+    raise NotImplementedError()
+
+
+def serialize_sklearn_estimator(estimator):
+    """Serialize model to JSON, see https://cmry.github.io/notes/serialize for inspiration."""
+    raise NotImplementedError()
+
+
+def deserialize_sklearn_estimator(js):
+    """Init model from JSON."""
+    raise NotImplementedError()
+
+
+def sklearn_to_pfa(estimator):
+    """Convert scikit-learn to PFA."""
+    # QUESTION: What's the best way for generating PFA code from scikit models in your opinion? Should I use PrettyPFA
```

Comment:
Can you try using PrettyPFA? It should work well from Python

Reply:
PrettyPFA is neat, building regression model was indeed trivial.

The only problem with PrettyPFA is that titus does not support python 3... We have 3 options here (in order of my preference):
1. Translate PrettyPFA to PFA in Woken, i.e. `io_helper.save_results(pfa, '', 'pfa_pretty')`
2. Use (unofficial) python 3 fork https://github.com/animator/hadrian and convert it in the container
3. Use python 2 and convert it in the container